README.md
# Kaizen Flow

Kaizen Flow is a robust web application designed to streamline workflow and productivity. Built with Vue 3, it incorporates a suite of tools and libraries to provide a seamless user experience.

## Table of Contents
- [Kaizen Flow](#kaizen-flow)
  - [Table of Contents](#table-of-contents)
  - [Introduction](#introduction)
  - [Author](#author)
  - [Scripts](#scripts)
    - [start](#start)
    - [web-app](#web-app)
    - [build](#build)
    - [server](#server)
    - [test:e2e](#teste2e)
  - [Dependencies](#dependencies)

## Introduction
<a id="introduction"></a>
Welcome to the "Kaizen Flow" project, a web application designed for efficiency and productivity. This project leverages modern web technologies to create a seamless and dynamic user experience.

## Author
<a id="author"></a>

## Scripts
<a id="scripts"></a>
Below are the scripts defined in the `package.json` file for various tasks required during development, testing, and deployment.

### start

<a id="start"></a>
Concurrently runs both the Vite development server and the Node.js server.
Navigate to the application directory open your terminal and run the following

```bash
npm install
npm i -g concurrently
npm run start
```

### web-app
<a id="web-app"></a>
Starts the Vite development server for the front-end application.
```bash
npm run web-app
```

### build

<a id="build"></a>
Runs the build-only script to create a production build of the application.

```bash
npm run build
```

### server

<a id="server"></a>
Launches the Node.js server found in the server/server.js file.

```bash
npm run server
```

### test:e2e

<a id="teste2e"></a>
Runs end-to-end tests by starting the server and frontend and then executing Cypress tests.

```bash
npm run test:e2e
```


## Dependencies

<a id="dependencies"></a>
A list of production dependencies required for the application to run. These include frameworks, libraries, and other tools used in the project.
DevDependencies


- [@fullcalendar/core](https://www.npmjs.com/package/@fullcalendar/core)
- [@fullcalendar/daygrid](https://www.npmjs.com/package/@fullcalendar/daygrid)
- [@fullcalendar/interaction](https://www.npmjs.com/package/@fullcalendar/interaction)
- [@fullcalendar/timegrid](https://www.npmjs.com/package/@fullcalendar/timegrid)
- [@fullcalendar/vue3](https://www.npmjs.com/package/@fullcalendar/vue3)
- [body-parser](https://www.npmjs.com/package/body-parser)
- [cors](https://www.npmjs.com/package/cors)
- [express](https://www.npmjs.com/package/express)
- [ics](https://www.npmjs.com/package/ics)
- [moment](https://www.npmjs.com/package/moment)
- [multer](https://www.npmjs.com/package/multer)
- [pdf-text-extract](https://www.npmjs.com/package/pdf-text-extract)
- [pinia](https://www.npmjs.com/package/pinia)
- [vue](https://www.npmjs.com/package/vue)
- [vue-router](https://www.npmjs.com/package/vue-router)


<a id="devdependencies"></a>
Development dependencies are required only during the development process, not in the production environment. This includes testing frameworks, build tools, and other utilities.
Contributing

- [@tsconfig/node18](https://www.npmjs.com/package/@tsconfig/node18)
- [@types/jsdom](https://www.npmjs.com/package/@types/jsdom)
- [@types/node](https://www.npmjs.com/package/@types/node)
- [@vitejs/plugin-vue](https://www.npmjs.com/package/@vitejs/plugin-vue)
- [@vue/test-utils](https://www.npmjs.com/package/@vue/test-utils)
- [@vue/tsconfig](https://www.npmjs.com/package/@vue/tsconfig)
- [autoprefixer](https://www.npmjs.com/package/autoprefixer)
- [concurrently](https://www.npmjs.com/package/concurrently)
- [cypress](https://www.npmjs.com/package/cypress)
- [jsdoc](https://www.npmjs.com/package/jsdoc)
- [jsdom](https://www.npmjs.com/package/jsdom)
- [kleur](https://www.npmjs.com/package/kleur)
- [npm-run-all](https://www.npmjs.com/package/npm-run-all)
- [postcss](https://www.npmjs.com/package/postcss)
- [sass](https://www.npmjs.com/package/sass)
- [start-server-and-test](https://www.npmjs.com/package/start-server-and-test)
- [tailwindcss](https://www.npmjs.com/package/tailwindcss)
- [typescript](https://www.npmjs.com/package/typescript)
- [vite](https://www.npmjs.com/package/vite)
- [vitest](https://www.npmjs.com/package/vitest)

<a id="contributing"></a>
Contributions are always welcome! If you're interested in contributing, please read the contributing guide and follow the project's code of conduct.




CHANGELOG.md
# CHANGELOG

## 📛 Version 1.1 - Innocent WK Chinyemba
- **Date:** Tue Nov 28 05:56:19 2023 +0200
    - rename CHANGE LOG file
    - A new name for a fresh perspective! 📛

## 🔄 Version 1.2 - Innocent WK Chinyemba
- **Date:** Tue Nov 28 05:30:55 2023 +0200
    - update phase 3 docs
    - Sprucing things up with update number 2! 🌟

## 🔄 Version 1.3 - Innocent WK Chinyemba
- **Date:** Tue Nov 28 05:30:13 2023 +0200
    - update docs
    - Sprucing things up with update number 3! 🌟

## 🗑️ Version 1.4 - Innocent WK Chinyemba
- **Date:** Tue Nov 28 04:45:46 2023 +0200
    - remove unused code
    - Cleaning house, removing what we don't need. 🧹

## 🗑️ Version 1.5 - Innocent WK Chinyemba
- **Date:** Tue Nov 28 04:28:25 2023 +0200
    - remove unused code
    - Cleaning house, removing what we don't need. 🧹

## 🔄 Version 1.6 - Innocent WK Chinyemba
- **Date:** Tue Nov 28 03:51:40 2023 +0200
    - phase 3 documentation update
    - Sprucing things up with update number 6! 🌟

## 🔄 Version 1.7 - Innocent WK Chinyemba
- **Date:** Mon Nov 27 22:08:56 2023 +0200
    - update README
    - Sprucing things up with update number 7! 🌟

## 🔄 Version 1.8 - Innocent WK Chinyemba
- **Date:** Mon Nov 27 06:22:47 2023 +0200
    - update readme
    - Sprucing things up with update number 8! 🌟

##  Version 1.9 - Innocent WK Chinyemba
- **Date:** Thu Nov 9 01:12:38 2023 +0200
    - change start date and JSON data object wrapper

## 🆕 Version 1.10 - Innocent WK Chinyemba
- **Date:** Thu Nov 9 00:18:36 2023 +0200
    - create docs
    - Creating something new, watch this space! 🌱

## ✨ Version 1.11 - Innocent WK Chinyemba
- **Date:** Thu Nov 9 00:12:04 2023 +0200
    - feat: Add cleanSubChapters module
    - Added a new module 'cleanSubChapters' which processes an array of unit
    - objects by cleaning up subChapterTitles to remove any trailing dots and
    - excess whitespace. The module also extracts and appends the pageNumber
    - property to each subChapter entry.
    - Adding new features to enhance our journey. ✨
    - Cleaning house, removing what we don't need. 🧹

##  Version 1.12 - Innocent WK Chinyemba
- **Date:** Wed Nov 8 23:28:58 2023 +0200
    - clean up units json object parsing

## 🔄 Version 1.13 - Innocent WK Chinyemba
- **Date:** Wed Nov 8 22:43:11 2023 +0200
    - version update
    - Sprucing things up with update number 13! 🌟

## 🗑️ Version 1.14 - Innocent WK Chinyemba
- **Date:** Wed Nov 8 00:08:25 2023 +0200
    - remove unnecessary default components and features
    - Cleaning house, removing what we don't need. 🧹

##  Version 1.15 - Innocent WK Chinyemba
- **Date:** Tue Nov 7 23:29:58 2023 +0200
    - change default branch to main

## 🔄 Version 1.16 - Innocent WK Chinyemba
- **Date:** Tue Nov 7 22:48:20 2023 +0200
    - update tests
    - Sprucing things up with update number 16! 🌟

## 🔧 Version 1.17 - Innocent WK Chinyemba
- **Date:** Tue Nov 7 21:40:45 2023 +0200
    - Fix path resolution for pdftotext windows binary
    - Tweaking and fixing, ironing out the kinks for a smoother ride. 🛠️

##  Version 1.18 - Innocent WK Chinyemba
- **Date:** Tue Nov 7 20:20:34 2023 +0200
    - resolve package conficts in test environment

##  Version 1.19 - Innocent WK Chinyemba
- **Date:** Mon Nov 6 01:22:29 2023 +0200
    - server packages

##  Version 1.20 - Innocent WK Chinyemba
- **Date:** Mon Nov 6 01:21:30 2023 +0200
    - Build: 06.11.2023 01:21

## ✨ Version 1.21 - Innocent WK Chinyemba
- **Date:** Mon Nov 6 01:20:17 2023 +0200
    - add dynamic routes to calender from upload
    - Adding new features to enhance our journey. ✨

## 🔀 Version 1.22 - Innocent WK Chinyemba
- **Date:** Mon Nov 6 00:30:32 2023 +0200
    - resolve merge conflict #2
    - Merging branches, bringing our efforts together. 🤝

##  Version 1.23 - Innocent WK Chinyemba
- **Date:** Mon Nov 6 00:26:54 2023 +0200
    - ignore doc files

##  Version 1.24 - Innocent WK Chinyemba
- **Date:** Mon Nov 6 00:26:26 2023 +0200
    - hide unused links

## ✨ Version 1.25 - Innocent WK Chinyemba
- **Date:** Mon Nov 6 00:26:13 2023 +0200
    - add calender creator button
    - Adding new features to enhance our journey. ✨

## ✨ Version 1.26 - Innocent WK Chinyemba
- **Date:** Mon Nov 6 00:06:02 2023 +0200
    - add selectable ics files
    - Adding new features to enhance our journey. ✨

##  Version 1.27 - Innocent WK Chinyemba
- **Date:** Sun Nov 5 21:22:42 2023 +0200
    - import moment timezone

## ✨ Version 1.28 - Innocent WK Chinyemba
- **Date:** Sun Nov 5 21:21:09 2023 +0200
    - Implemented unified resource retrieval endpoint
    - - Created a new /get-resources endpoint in the Express router.
    - - Enabled retrieval of both 'modules' and 'calendars' via a required 'type' query parameter.
    - - Allowed fetching of specific files when the 'file' query parameter is provided.
    - - Included input validation for the 'type' parameter to ensure it only accepts 'modules' or 'calendars'.
    - - Added error handling for file existence checks and directory reading failures.
    - - Modified error messages to provide clearer descriptions and, where applicable, error details.
    - Adding new features to enhance our journey. ✨
    - Creating something new, watch this space! 🌱

## 🆕 Version 1.29 - Innocent WK Chinyemba
- **Date:** Sun Nov 5 21:17:17 2023 +0200
    - create ics parser
    - Creating something new, watch this space! 🌱

## 🛠️ Version 1.30 - Innocent WK Chinyemba
- **Date:** Sun Nov 5 18:44:43 2023 +0200
    - Refactor generate-ics endpoint to use fs.promises for async file operations
    - Refactoring for efficiency, because neatness counts! 📐

##  Version 1.31 - Innocent WK Chinyemba
- **Date:** Sun Nov 5 18:43:44 2023 +0200
    - customise calender styling

##  Version 1.32 - Innocent WK Chinyemba
- **Date:** Sun Nov 5 17:51:15 2023 +0200
    - ics creator api path

## ✨ Version 1.33 - Innocent WK Chinyemba
- **Date:** Sun Nov 5 15:45:04 2023 +0200
    - add calender view
    - Adding new features to enhance our journey. ✨


server/lib/ics.js
const { writeFileSync } = require('fs');
const ics = require('ics');

const data = {
  "units": {
    "units": [
      {
        "unitNumber": "Unit 1",
        "subChapters": [
          {
            "subChapterNumber": "1.1",
            "subChapterTitle": "IT Services . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 15"
          },
          {
            "subChapterNumber": "1.2",
            "subChapterTitle": "IT Service Management . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 17"
          },
          {
            "subChapterNumber": "1.3",
            "subChapterTitle": "ITSM Frameworks . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 19"
          }
        ]
      },
      {
        "unitNumber": "Unit 2",
        "subChapters": [
          {
            "subChapterNumber": "2.1",
            "subChapterTitle": "Stakeholders, Services, and Service Management . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 24"
          },
          {
            "subChapterNumber": "2.2",
            "subChapterTitle": "Value Contribution of IT . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 28"
          },
          {
            "subChapterNumber": "2.3",
            "subChapterTitle": "Four Dimensions Model . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 29"
          }
        ]
      },
      {
        "unitNumber": "Unit 3",
        "subChapters": [
          {
            "subChapterNumber": "3.1",
            "subChapterTitle": "Basics and Overview . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 34"
          },
          {
            "subChapterNumber": "3.2",
            "subChapterTitle": "Inputs, Outcome, and Governance . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 36"
          },
          {
            "subChapterNumber": "3.3",
            "subChapterTitle": "The Service Value Chain (SVC) . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 37"
          },
          {
            "subChapterNumber": "3.4",
            "subChapterTitle": "Continual Improvement . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 39"
          }
        ]
      },
      {
        "unitNumber": "Unit 4",
        "subChapters": [
          {
            "subChapterNumber": "4.1",
            "subChapterTitle": "Overview . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 44"
          },
          {
            "subChapterNumber": "4.2",
            "subChapterTitle": "Value Orientation . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 46"
          },
          {
            "subChapterNumber": "4.3",
            "subChapterTitle": "Iterative Procedure and Feedback . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 49"
          },
          {
            "subChapterNumber": "4.4",
            "subChapterTitle": "Establish Collaboration and Visibility . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 51"
          },
          {
            "subChapterNumber": "4.5",
            "subChapterTitle": "Optimize and Automate . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 54"
          }
        ]
      },
      {
        "unitNumber": "Unit 5",
        "subChapters": [
          {
            "subChapterNumber": "5.1",
            "subChapterTitle": "Overview . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 58"
          },
          {
            "subChapterNumber": "5.2",
            "subChapterTitle": "General Management Practices . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 61"
          },
          {
            "subChapterNumber": "5.3",
            "subChapterTitle": "Service Management Practices . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 64"
          },
          {
            "subChapterNumber": "5.4",
            "subChapterTitle": "Technical Practices . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 68"
          }
        ]
      },
      {
        "unitNumber": "Unit 6",
        "subChapters": [
          {
            "subChapterNumber": "6.1",
            "subChapterTitle": "Structure and Elements of BSI IT Basic Protection Framework . . . . . . . . . . . . . . . . . . . . 72"
          },
          {
            "subChapterNumber": "6.2",
            "subChapterTitle": "Information Security Process . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 77"
          }
        ]
      }
    ]
  }
};

// Total available hours for the schedule
const totalHours = 150;

// Start date for the schedule (adjust as needed)
const startDate = new Date(2023, 11, 15, 8, 0, 0); // December 15, 2023, 08:00 AM

// Initialize the schedule
const schedule = [];

// Calculate the distribution of hours
const topics = data.units.units;
const topicHours = totalHours / topics.length;

// Define the time window to avoid (23:00 to 05:00)
const avoidStartTime = 23;
const avoidEndTime = 5;

// Define the maximum study duration (in hours) before taking a break
const maxStudyDuration = 1;

// Loop through each topic
for (const topic of topics) {
  const subChapters = topic.subChapters;
  const subChapterHours = topicHours / subChapters.length;

  // Loop through sub-chapters within the topic
  for (const subChapter of subChapters) {
    // Calculate start and end times for the sub-chapter
    const startTime = new Date(startDate);
    const endTime = new Date(startTime.getTime() + subChapterHours * 60 * 60 * 1000);

    // Avoid the time window (23:00 to 05:00)
    if (startTime.getHours() >= avoidStartTime || endTime.getHours() <= avoidEndTime) {
      startTime.setHours(avoidEndTime);
      endTime.setHours(avoidEndTime);
    }

    // Create an event for the sub-chapter
    const event = {
      title: `Unit ${topic.unitNumber} - Subchapter ${subChapter.subChapterNumber}`,
      description: subChapter.subChapterTitle,
      start: [startTime.getFullYear(), startTime.getMonth() + 1, startTime.getDate(), startTime.getHours(), startTime.getMinutes()],
      duration: { hours: subChapterHours },
    };

    // Add the event to the schedule
    schedule.push(event);

    // Update the start time for the next sub-chapter
    startDate.setTime(endTime.getTime());

    // Calculate the break time
    const breakDuration = maxStudyDuration - subChapterHours;
    if (breakDuration > 0) {
      // Add a break event
      const breakStartTime = new Date(startDate);
      const breakEndTime = new Date(startDate.getTime() + breakDuration * 60 * 60 * 1000);

      // Avoid the time window (23:00 to 05:00)
      if (breakStartTime.getHours() >= avoidStartTime || breakEndTime.getHours() <= avoidEndTime) {
        breakStartTime.setHours(avoidEndTime);
        breakEndTime.setHours(avoidEndTime);
      }

      const breakEvent = {
        title: 'Break',
        description: 'Take a break',
        start: [breakStartTime.getFullYear(), breakStartTime.getMonth() + 1, breakStartTime.getDate(), breakStartTime.getHours(), breakStartTime.getMinutes()],
        duration: { hours: breakDuration },
      };

      // Add the break event to the schedule
      schedule.push(breakEvent);

      // Update the start time for the next sub-chapter or break
      startDate.setTime(breakEndTime.getTime());
    }
  }
}

// Generate the ICS content
const { error, value } = ics.createEvents(schedule);

if (error) {
  console.error(error);
} else {
  // Save the ICS file to disk
  writeFileSync(`${__dirname}/schedule.ics`, value);
  console.log('Schedule saved as schedule.ics');
}
